openstack/client_test.go

```diff
@@ -159,3 +159,95 @@ func TestAuthenticatedClientV2(t *testing.T) {
 }
+
+func TestNewClient(t *testing.T) {
+	client, err := NewClient("https://example.com")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/v2.0")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/v2.0/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/v2.0/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/v2.0/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/foo/bar")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/foo/bar/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/foo/bar/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/foo/bar/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/foo/bar/v2.0")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/foo/bar/v2.0/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/foo/bar/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/foo/bar/v2.0/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/foo/bar/v2.0/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/foo/bar/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/foo/bar/v3/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/foo/bar/v3/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/foo/bar/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/v3")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/v3/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	client, err = NewClient("https://example.com/v3/")
+	th.AssertNoErr(t, err)
+	th.AssertEquals(t, "https://example.com/v3/", client.IdentityEndpoint)
+	th.AssertEquals(t, "https://example.com/", client.IdentityBase)
+
+	_, err = NewClient("https://example.com/v2.3/")
+	th.AssertErr(t, err)
+	expected := fmt.Errorf("Invalid identity endpoint version %v. Supported versions: v2.0, v3", "v2.3")
+	th.AssertEquals(t, expected.Error(), err.Error())
+
+	_, err = NewClient("https://example.com/v2.0/foo")
+	th.AssertErr(t, err)
+	expected = fmt.Errorf("Path suffixes (after version) are not supported.")
+	th.AssertEquals(t, expected.Error(), err.Error())
+
+	// Does not match regexp, include to base
+	client, err = NewClient("https://example.com/v2a0/")
```

Comment:
In my opinion `/v2a0/`is a non-version path, and should be treated like `/foo/`. Now the question is, does the endpoint have to contain version, or should we treat the path as IdentityBase in case it doesn't. In the current proposal versionless endpoints are accepted and assigned as IdentityBase with an empty IdentityEndpoint.